text_to_num/lang/portuguese.py

```diff
@@ -0,0 +1,348 @@
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import re
+import sys
+from typing import Dict, Optional, Set, Tuple, List, Union
+
+from .base import Language
+
+#
+# CONSTANTS
+# Built once on import.
+#
+
+# Those words multiplies lesser numbers (see Rules)
+# Exception: "(de) milliards" that can multiply bigger numbers ("milliards de milliards")
+MULTIPLIERS = {
+    "mil":      1000,
+    "milhar":   1000,
+    "milhares": 1000,
+    "milhão":   1000000,
+    "milhao":   1000000,
+    "milhões":  1000000,
+}
+
+
+# Units are terminals (see Rules)
+UNITS: Dict[str, int] = {
+    word: value
+    for value, word in enumerate(
+        "um dois três quatro cinco seis sete oito nove".split(), 1
+    )
+}
+# Unit variants
+UNITS["uma"] = 1
+UNITS["duas"] = 2
+
+# Single tens are terminals (see Rules)
+# exact find
+STENS: Dict[str, int] = {
+    word: value
+    for value, word in enumerate(
+        "dez onze doze treze quatorze quinze dezasseis dezassete dezoito dezanove"
+        .split(),
+        10,
+    )
+}
+
+# Ten multiples
+# Ten multiples may be followed by a unit only;
+# the number is the multiplier of the first token
+MTENS: Dict[str, int] = {
+    word: value * 10
+    for value, word in enumerate(
+        "vinte trinta quarenta cinquenta sessenta setenta oitenta noventa".split(), 2
+    )
+}
+
+# Ten multiples that can be combined with STENS
+MTENS_WSTENS: Set[str] = set()
+
+HUNDRED = {
+    "cem": 100,
+    "centena": 100,
+    "cento": 100,
+    "centenas": 100,
+    "duzentos": 200,
+    "duzentas": 200,
+    "trezentos": 300,
+    "trezentas": 300,
+    "quatrocentos": 400,
+    "quatrocentas": 400,
+    "quinhentos": 500,
+    "quinhentas": 500,
+    "seiscentos": 600,
+    "seiscentas": 600,
+    "setecentos": 700,
+    "setecentas": 700,
+    "oitocentos": 800,
+    "oitocentas": 800,
+    "novecentos": 900,
+    "novecentas": 900,
+}
+
+# Composites are tens already composed with terminals in one word.
+# Composites are terminals.
+
+COMPOSITES: Dict[str, int] = {}
+
+# All number words
+NUMBERS = MULTIPLIERS.copy()
+NUMBERS.update(UNITS)
+NUMBERS.update(STENS)
+NUMBERS.update(MTENS)
+NUMBERS.update(HUNDRED)
+NUMBERS.update(COMPOSITES)
+
+class Portuguese(Language):
+
+    ISO_CODE = 'pt'
+
+    MULTIPLIERS = MULTIPLIERS
+    UNITS = UNITS
+    STENS = STENS
+    MTENS = MTENS
+    MTENS_WSTENS = MTENS_WSTENS
+    HUNDRED = HUNDRED
+    NUMBERS = NUMBERS
+
+    SIGN = {"mais": "+", "menos": "-"}
+    ZERO = {"zero"}
+    DECIMAL_SEP = "vírgula"
+    DECIMAL_SYM = ","
+
+    DIRECT_CONVERSION = UNITS.copy()
+    DIRECT_CONVERSION.update(STENS)
+    DIRECT_CONVERSION = { k : str(v) for k, v in DIRECT_CONVERSION.items()}
```

Comment:
Hi Luis, thanks for the patch. You have type issue here because DIRECT_CONVERSION type is copied from UNITS

           text_to_num/lang/portuguese.py:168: error: Argument 1 to "update" of "dict" has incompatible type "Dict[str, str]"; expected "Mapping[str, int]"

You'd better use ints as a values in a dictionary. Try with mypy.

Reply:
solved, it was actually a leftover var.